### DISCUSSION TITLE
Hi, Need help, sorry, what does this mean?! "Uninitialized "output" in Deployer container."

### DISCUSSION THREAD

**Discussion by rmpel** [deployphp/deployer#3966] [Help needed]
Hi, Need help, sorry, what does this mean?! 

`deployer --version`
``` response
 RuntimeException  in Deployer.php on line 105:
  Uninitialized "output" in Deployer container.
```

How do I combat this? where do I start debugging?

---

Background information:

we're trying to upgrade from Dep6 to Dep7
We followed the upgrade instructions, and locally, we ge no errors.
But when executing in a BitBucket Pipeline, even with just trying to print the version number, we get this error.
I don't understand the error, so I don't know what we're doing wrong.

Deployer version used: 7.5.5
PHP 8.1 (project requirement) as well as PHP 8.3 (local testing)

![Screenshot 2024-11-26 at 12 34 31](https://github.com/user-attachments/assets/15315ac3-ed58-4fd4-9d1e-636c7c832b7e)

**Comment by antonmedv** [comment]
Please show whole output with -vvv.

**Comment by rmpel** [comment]
```
+ deployer --version -vvv
 RuntimeException  in Deployer.php on line 105:
  Uninitialized "output" in Deployer container.
#0 phar:///usr/local/bin/deployer/src/Component/Pimple/Container.php(124): Deployer\Deployer->Deployer\{closure}()
#1 phar:///usr/local/bin/deployer/src/Deployer.php(248): Deployer\Component\Pimple\Container->offsetGet()
#2 phar:///usr/local/bin/deployer/src/functions.php(663): Deployer\Deployer->__get()
#3 /opt/atlassian/pipelines/agent/build/deploy-7x.php(66): Deployer\warning()
#4 /opt/atlassian/pipelines/agent/build/deploy.php(11): require('...')
#5 phar:///usr/local/bin/deployer/src/Importer/Importer.php(66): require('...')
#6 [internal function]: Deployer\Importer\Importer::Deployer\Importer\{closure}()
#7 phar:///usr/local/bin/deployer/src/Importer/Importer.php(75): call_user_func()
#8 phar:///usr/local/bin/deployer/src/Deployer.php(314): Deployer\Importer\Importer::import()
#9 phar:///usr/local/bin/deployer/bin/dep(96): Deployer\Deployer::run()
#10 /usr/local/bin/deployer(4): require('...')
#11 {main}
```

with thanks for your response :)

**Comment by rmpel** [comment]
I suspect this is the issue;
```/opt/atlassian/pipelines/agent/build/deploy-7x.php(66): Deployer\warning()```
but why? Why can I not output warnings?

**Comment by rmpel** [comment]
```
function warning(string $message): void
{
    $message = "<fg=yellow;options=bold>warning</> <comment>$message</comment>";

    if (Context::has()) {
        writeln($message);
    } else {
        Deployer::get()->output->writeln($message);
    }
}
```
The source code suggests that it should be possible to `warning` (or `info`) out of context (a host context), so why do I have to jump through hoops like this;

```
<?php

/**
 * Writes an info message, without a context.
 *
 * @param string $message
 *
 * @return void
 */
function info_no_context( string $message ): void {
	// Print with ANSI Escape colors.
	echo "\033[32m✔ $message \033[0m \n";
}

/**
 * Writes a warning message, without a context.
 *
 * @param string $message
 *
 * @return void
 */
function warning_no_context( string $message ): void {
	echo "\033[33m⚠ $message \033[0m \n";
}
```

I'm not complaining, I just want to understand :)

**Reply by antonmedv** [reply]
Cloud you please share your code for deploy.php?

**Reply by rmpel** [reply]
I'm going to put this on hold for now as I feel this will take much much more time to upgrade than we had expected. And I don't want to burden you with my ramblings. I had them all typed up, but it would serve no-one, so I removed it.

I appreciate your attempts and effort and especially the project as a whole, and I hope Deployer 6 will stay available for the near and far future because I don't see a happy upgrade path here...

**Reply by antonmedv** [reply]
Thanks. I keep all versions in here: https://deployer.org/download
